---
title: "QTL schematic plots"
author: "Jack Humphrey"
date: "18/12/2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggbio)
library(GenomicRanges)
library(ggrepel)
library(ggtranscript)
library(patchwork)

theme_jh <- function () { 
    theme_bw(base_size=5, base_family="Helvetica") %+replace% 
        theme(
          panel.grid = element_blank(),
          strip.background = element_blank(),
          #panel.border = element_blank(),
          axis.line = element_line(),
          axis.ticks = element_line(colour = "black"),
          #text = element_text(color = "black"), 
          strip.text = element_text(color = "black"),
          axis.text = element_text(colour = "black"),
            panel.background  = element_blank(),
            plot.background = element_rect(fill="white", colour=NA), 
            legend.background = element_rect(fill="transparent", colour=NA),
            legend.key = element_rect(fill="transparent", colour=NA), legend.text = element_text(size = 7)
        )
}
```

For a selection of loci:

Plot GWAS and QTL SNP positions
Plot transcripts, identified by COLOC
Plot junctions



```{r}
data_file <-  here::here("COLOC/mmQTL/v5/coloc_schematic_plot_data.RData")
if(!file.exists(data_file)){

sqanti <- read_tsv(here::here("data/2023_stringtie_mix/raj_roussos_0922_stringtie_mix_filter_sqanti_classification.tsv.gz"))
gene_meta <- read_tsv(here::here("external/gencode.v38.primary_assembly.tx2gene.tsv.gz"))

sqanti$gene <- gene_meta$gene_name[match(sqanti$associated_gene, gene_meta$gene_id)]

ca_meta <- read_csv(here::here("external/Rousso_caQTLs/PeakInfoDetailed_RK_11_25_20.csv.gz"))

#finemap_res <- data.table::fread(here::here("data/Fine_Mapping/brian/microgliaQTL_multiGWAS.finemapping_merged.labelled.tsv"))

#data.class <- read.table(class.file, header=T, as.is=T, sep="\t")

#union_gff <- rtracklayer::import("~/GENCODE/gencode.v30.annotation.gtf.gz")
#isoseq_gff <- rtracklayer::import(here::here("data/post_filter/raj_roussos_miss_sqanti.cds.gtf.sorted.gtf.gz"))

#isoseq_gff$gene_name <- all_coloc$gene_name[ match(isoseq_gff$gene_id, all_coloc$gene)]

#isoseq_gff <- isoseq_gff[ isoseq_gff$gene_name %in% all_coloc$gene_name]

# get out novel transcripts
#isoseq_gff <- isoseq_gff[ grepl("tx.", isoseq_gff$transcript_id ) & isoseq_gff$gene_name %in% all_coloc$gene_name ]

## IMPORT GENCODE
#union_gff <- rtracklayer::import(here::here("~/GENCODE/gencode.v38.primary_assembly.annotation.gtf.gz"))
# select genes used for COLOC
#union_gff <- union_gff[ union_gff$gene_name %in% all_coloc$gene_name]

# Import GENCODE+Novel
union_gff <- rtracklayer::import(here::here("data/2023_stringtie_mix/raj_roussos_0922_stringtie_mix_combined.sorted.gtf.gz") )

all_coloc <- read_tsv(here::here("COLOC/mmQTL/all_coloc_results_annotated.tsv")) %>%
  filter(distance_filter == "PASS")

union_gff <- union_gff[ union_gff$gene_name %in% all_coloc$QTL_Gene | union_gff$transcript_id %in% all_coloc$QTL_Gene]

## ashvin's fine-mapping results
finemap_res <- read_csv(here::here("external/Ashvin_fine_mapping/Bellenguez_LD_Results.csv")) %>% janitor::clean_names()

finemap_bed <- finemap_res %>% 
  mutate(chr = paste0("chr", chr), 
         start = bp - 1,
         coord = paste0(chr, ":", bp)
         ) %>%
  mutate(name = coalesce(rsid, coord))


# write_tsv(  select(finemap_bed, chr, start, bp, name),
#             here::here("external/Ashvin_fine_mapping/Bellenguez_LD_Results.bed"),col_names = FALSE
# )
# 
# write_tsv(  select(finemap_bed, chr, start, bp, pip),
#             here::here("external/Ashvin_fine_mapping/Bellenguez_LD_Results.bedGraph"),col_names = FALSE
# )
save.image(data_file)
}else{
  load(data_file)
}




```

Plotting functions

```{r}

finemap_plot <- function(mylocus, col_width = 100, label = FALSE, min_pip = 0.05, xlim = NULL){
  stopifnot(mylocus %in% finemap_bed$locus)
  finemap_loc <- filter(finemap_bed, locus == mylocus) %>%
    mutate(name = ifelse(pip > min_pip, name, "")) 
  
  lead_snp <- arrange(finemap_loc, p) %>% head(1)
  
  if( !is.null(xlim)){
    finemap_loc <- filter(finemap_loc, bp > xlim[1] & bp < xlim[2] )
  }
  
  # make plot
  p <- finemap_loc %>%
    mutate(credible_set = factor(credible_set)) %>%
   ggplot(aes(x= bp, y = pip) ) + 
    geom_col( aes(colour =credible_set, fill = credible_set), width = col_width, position = position_dodge(width = 2*col_width) ) +
    theme_classic() +
    labs(colour = "credible set", fill = "credible set", x = "position") +
    geom_point(data = lead_snp, aes(x = bp, y = 1), colour = "black", fill = "purple", shape = 23) +
    scale_y_continuous(expand = c(0,0), limits = c(0,1.1), name = "PIP") +
    scale_fill_viridis_d() + scale_colour_viridis_d()  +
    theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=10))
  
  if( !is.null(xlim) ){
    p <- p + scale_x_continuous(limits = xlim)
  }
  if(label == TRUE){
   p <- p + geom_text_repel(aes(label = name ), size = 3, nudge_y = 0.025, min.segment.length = unit(0, units = "lines"))
  }
  return(p)
}

finemap_plot("PLCG2-1", 100, min_pip = 0.2, label = TRUE)#, xlim = c(81800000, 81900000) )
finemap_plot("TREM2-2", 10)


# plot transcripts
isoform_plot <- function(mylocus, mygene = NULL, PP4 = 0.8, transcript_type = "isoseq", level = "expression", mylim = c(-0.6,0.6), gwas = "Bellenguez_2021", guide = TRUE, colourby = "QTL_Beta" ){
  if( transcript_type == 'isoseq'){ transcript_type <- "GENCODE+Novel"}
  
  all_features <- filter(all_coloc, locus == mylocus, PP.H4.abf > PP4, GWAS == gwas  ) %>%
    pull(gene)
  
  coloc_loc <- 
    filter(all_coloc, locus == mylocus, gene %in% all_features, GWAS == gwas  ) %>%
    arrange(PP.H4.abf) %>%
    filter(set == level, reference == transcript_type)
  
  print(coloc_loc)
  
  if( !is.null(mygene)){
    coloc_loc <- filter(coloc_loc, gene_name %in% mygene )
  }
  
  stopifnot( nrow(coloc_loc) > 0)
  if(level == "transcript"){
    gff_loc <- union_gff[ union_gff$transcript_id %in% coloc_loc$QTL_Gene & 
                             union_gff$type %in% c("exon", "CDS")]
    gff_loc$QTL_Beta <- coloc_loc$GWAS_SNP_Beta[ match(gff_loc$transcript_id, coloc_loc$QTL_Gene)]
    
    gff_loc$transcript_id <- factor(gff_loc$transcript_id, levels = coloc_loc$QTL_Gene )
    print(gff_loc)
    groupby <- "transcript_id"
  }
  if(level == "expression"){
    gff_loc <- union_gff[ union_gff$gene_name %in% coloc_loc$gene_name & 
                             union_gff$type %in% c("exon", "CDS")]
    stopifnot(length(gff_loc) > 0 )
    gff_loc$gene_name <- factor(gff_loc$gene_name)
    gff_loc$QTL_Beta <- coloc_loc$GWAS_SNP_Beta[ match(gff_loc$gene_name, coloc_loc$gene_name)]
    groupby <- "gene_name"
  }
  # NEW - ggtranscript!
  exons <- as.data.frame(gff_loc[ gff_loc$type == "exon"])
  cds <- as.data.frame(gff_loc[gff_loc$type == "CDS"])
  plot <- 
    exons %>% 
      ggplot(aes_string(
        xstart = "start",
        xend = "end",
        y = groupby,
        fill = colourby
    )) +
    geom_intron(arrow.min.intron.length = 5000,
        data = to_intron(as.data.frame(gff_loc), "transcript_name"),
        aes(strand = strand)
    ) +
    geom_range(height = 0.25) +
    geom_range(data = cds, height = 0.5) +
    labs(y = "") +
    scale_fill_distiller(name = colourby, limits = mylim, palette = "RdBu") + 
  # plot <- ggplot() +  
  #   geom_alignment(gff_loc,range.geom = "rect", gap.geom = "arrow", which = NULL,
  #                  aes_string(group = groupby, colour = "QTL_Beta", fill = "QTL_Beta"), colour = NA, 
  #                  truncate.gaps = TRUE) + 
    theme_classic() +
    theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=10), axis.text = element_text(colour = "black"))
  
  # if(colourby == "QTL_Beta"){
  #   plot <- plot + scale_fill_distiller(name = colourby, limits = mylim, palette = "RdBu")
  # }
  
  if(guide == FALSE){
    plot <- plot + guides(colour = "none", fill = "none")
  }
  
  return(plot)
}

peak_plot <- function(mylocus, PP4 = 0.8, gwas = "Bellenguez_2021"){
  coloc_loc <- 
    filter(all_coloc, locus == mylocus, PP.H4.abf >= PP4, GWAS == gwas  ) %>%
    arrange(PP.H4.abf) %>%
    filter( set_label == "chromatin accessibility") %>% 
    left_join(ca_meta, by = c("feature" = "PeakID"))
  
    ## Plot a bar
    grange_loc <-  
      GRanges(seqnames = coloc_loc$Chr, 
              ranges = IRanges(start = coloc_loc$start, end = coloc_loc$end), 
              peak = coloc_loc$feature, label = coloc_loc$feature, PP4 = coloc_loc$PP.H4.abf )
  grange_loc$fillby <- 0
  plot <- 
    ggplot() + 
    geom_alignment(data = grange_loc, aes(group = label, fill = PP4), range.geom = "rect" ) +
    theme_classic() + scale_fill_viridis_c() +
    theme(axis.text = element_text(colour = "black") ) + 
    labs(fill = "")
  return(plot)
  
}

peak_plot("PLCG2", PP4 = 0.1)

# 
isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.9)
# 
# n <- 5
# gwas <- "Bellenguez_2021"
# mygene <- "PLCG2"
# colourby <- "QTL_Beta"
# legend_pos <- "right"

# new separate function for leafcutter using ggtranscript
lc_plot <- function(mylocus, type = "leafcutter", gwas = "Bellenguez_2021", n = Inf, colourby = "QTL_Beta",
           legend_pos = "right", label = FALSE, PP4 = 0.8, mylim = c(-0.6,0.6), guide = TRUE, flank = 50){
    junctions_loc <- 
    filter(all_coloc, locus == mylocus, set == "leafcutter", GWAS == gwas, PP.H4.abf > PP4) %>%
      select(QTL_junction, feature, GWAS_SNP_Beta) %>%
    tidyr::separate(QTL_junction, into = c("chr", "start", "end"), sep = ":|-", remove = FALSE, convert = TRUE ) %>%
    #mutate(coord1 = paste0( as.numeric(start) - flank, "-", start),
    #       coord2 = paste0( end, "-", as.numeric(end) + flank)) %>%
    #  select(-start, -end) %>%
    #pivot_longer(names_to = "set",values_to = "coord", -c(QTL_junction, chr, feature,GWAS_SNP_Beta) ) %>%
    #mutate(start = as.numeric(str_split_fixed(coord, "-", 2)[,1]),
    #         end = as.numeric(str_split_fixed(coord, "-", 2)[,2]) ) %>%
    mutate(strand = ifelse(grepl("\\+", feature), "+", "-"))
    #mutate(start = as.numeric(start), end = as.numeric(end)) %>%
    #arrange(desc(PP.H4.abf)) %>%
    #head(n)
    print(junctions_loc)
    
    grange_loc <-  
    GRanges(seqnames = junctions_loc$chr, 
            ranges = IRanges(start = junctions_loc$start, end = junctions_loc$end), 
            strand = junctions_loc$strand,
            #PP.H4.abf = junctions_loc$PP.H4.abf, 
            #SNP = junctions_loc$QTL_SNP, 
            junction_id = junctions_loc$QTL_junction, 
            QTL_Beta = junctions_loc$GWAS_SNP_Beta)
  
  
  grange_loc$junction_id <- factor(grange_loc$junction_id)
  
  #return(grange_loc)
  grange_loc %>%
      ggplot(aes(
        xstart = start,
        xend = end,
        y = junction_id
    )) +
    geom_junction(junction.y.max = 0.5, size = 2,junction.orientation = "top",
        #data = to_intron(sod1_exons, "transcript_name"),
        aes(colour = QTL_Beta)
    ) + 
    labs(y = "") +
    scale_colour_distiller(name = "QTL Beta", limits = mylim, palette = "RdBu") +
          scale_fill_distiller(name = "QTL Beta", limits = mylim, palette = "RdBu") +
    theme_classic() + 
    theme(
          axis.text = element_text(colour = "black"), 
          legend.key.size = unit(0.25, 'cm'), 
          legend.title = element_text(size=10) 
          ) 
}

lc_plot("PLCG2", mylim = c(-1,1))

lc_plot("CD33", gwas = "Lambert_2013", mylim = c(-1,1))



# plot SUPPA QTLs
junction_plot <- 
  function(mylocus, type = "SE", gwas = "Bellenguez_2021", n = Inf, colourby = "QTL_Beta",
           legend_pos = "right", label = FALSE, PP4 = 0.8, mylim = c(-0.6,0.6), guide = TRUE, flank = 50){
    junctions_loc <- 
      filter(all_coloc, locus == mylocus, set == type, GWAS == gwas, PP.H4.abf > PP4)
    
    if( type %in% c("AF", "AL")){
      suppa_df <- str_split_fixed(junctions_loc$feature, pattern = ":|-", n = 10)
      strand <- unique(suppa_df[,10])
      #suppa_df[,2] <- paste0(suppa_df[,2], " ", 1:nrow(junctions_loc))
      # SUPPA ordering is mirrored between AF and AL with + and -
      if( (strand == "+" & type == "AF") | (strand == "-" & type == "AL") ){
        e1_start <- suppa_df[,4]
        e1_end <- suppa_df[,5]
        e2_start <- suppa_df[,7]
        e2_end <- suppa_df[,8]
        e3_start <- suppa_df[,6]
        e3_end <- as.numeric(e3_start) + flank
        e1_coord <- paste0(e1_start, "-", e1_end)
        e2_coord <- paste0(e2_start, "-", e2_end)
        e3_coord <- paste0(e3_start, "-", e3_end)
        event_id <- paste0(suppa_df[,2], ":", suppa_df[,4], "-", suppa_df[,9])
      }else{
        # AF - / AL +
        #start_coord <- 4; end_coord <- 6
        e1_end <- as.numeric(suppa_df[,4])
        e2_start <- suppa_df[,5]
        e2_end <- suppa_df[,6]
        e3_start <- suppa_df[,8]
        e3_end <- suppa_df[,9]
        e1_coord <- paste0(e1_end - flank, "-", e1_end)
        e2_coord <- paste0(e2_start, "-", e2_end)
        e3_coord <- paste0(e3_start, "-", e3_end)
        event_id <- paste0(suppa_df[,2], ":", suppa_df[,4], "-", suppa_df[,9])
      }
      coord_df <- 
        tibble(feature = junctions_loc$feature, QTL_junction = event_id, chr = suppa_df[,3],
               coord1 = e1_coord, coord2 = e2_coord, coord3 = e3_coord ) %>%
        pivot_longer(names_to = "set",values_to = "coord", -c(QTL_junction, chr, feature) ) %>%
        mutate(start = as.numeric(str_split_fixed(coord, "-", 2)[, 1]),
               end = as.numeric(str_split_fixed(coord, "-", 2)[,2]) ) #%>%
      #mutate(QTL_junction = paste0(QTL_junction, ":", coord))
      print(coord_df)
    }
    if( type %in% c("A3", "A5")){
      suppa_df <- str_split_fixed(junctions_loc$feature, pattern = ":|-", n = 8)
      strand <- unique(suppa_df[,8])
      if( (type == "A5" & strand == "+" ) | (type == "A3" & strand == "-") ){
        e1 <- paste0(as.numeric(suppa_df[,6]) - flank, "-", as.numeric(suppa_df[,6]) )
        e2 <- paste0(as.numeric(suppa_df[,4]) - flank, "-", as.numeric(suppa_df[,4]) )
        e3 <- paste0(as.numeric(suppa_df[,7]), "-", as.numeric(suppa_df[,7]) + flank )
        id <- paste0(suppa_df[,2], ":", suppa_df[,6], "-", suppa_df[,5])
      }else{
        e1 <- paste0(as.numeric(suppa_df[,4]) - flank, "-", as.numeric(suppa_df[,4]) )
        e2 <- paste0(as.numeric(suppa_df[,5]), "-", as.numeric(suppa_df[,5]) + flank )
        e3 <- paste0(as.numeric(suppa_df[,7]), "-", as.numeric(suppa_df[,7]) + flank )
        id <- paste0(suppa_df[,2], ":", suppa_df[,4], "-", suppa_df[,7])
      }
      coord_df <- 
        tibble(feature = junctions_loc$feature, QTL_junction = id, chr = suppa_df[,3],
               coord1 = e1, coord2 = e2, coord3 = e3) %>%
        pivot_longer(names_to = "set",values_to = "coord", -c(QTL_junction, chr, feature) ) %>%
        mutate(start = as.numeric(str_split_fixed(coord, "-", 2)[, 1]),
               end = as.numeric(str_split_fixed(coord, "-", 2)[,2]) )
      
    }
    if( type %in% c("SE")){
      # skipped exons - plot central exon plus surrounding introns with a little flanking sequence
      suppa_df <- str_split_fixed(junctions_loc$feature, pattern = ":", n = 6)
      coord1 <- suppa_df[,4]
      coord2 <- suppa_df[,5]
      exon_start <- as.numeric( gsub(".*-", "", coord1))
      exon_end <- as.numeric( gsub("-.*", "", coord2))
      exon_coord <- paste0( exon_start, "-", exon_end)
      flank_start <- as.numeric( gsub("-.*", "", coord1))
      flank_end <-  as.numeric( gsub(".*-", "", coord2))
      
      coord_df <- tibble(
        feature = junctions_loc$feature, QTL_junction = paste0(suppa_df[,2], ":", exon_coord), chr = suppa_df[,3], 
        coord1 = paste0(flank_start - flank, "-", flank_start), 
        coord2 = exon_coord, 
        coord3 = paste0(flank_end, "-", flank_end + flank)
      ) %>%
        pivot_longer(names_to = "set",values_to = "coord", -c(QTL_junction, chr, feature) ) %>%
        mutate(start = as.numeric(str_split_fixed(coord, "-", 2)[, 1]),
               end = as.numeric(str_split_fixed(coord, "-", 2)[,2]) ) #%>%
      # mutate(QTL_junction = paste0(QTL_junction, ":", coord))
      
    }
    if( type %in% c("RI") ){
      suppa_df <- str_split_fixed(junctions_loc$feature, pattern = ":", n = 7)
      coord_df <- 
        tibble(feature = junctions_loc$feature, QTL_junction = suppa_df[,2], chr = suppa_df[,3],
               coord = suppa_df[,5] ) %>%
        mutate(start = as.numeric(str_split_fixed(coord, "-", 2)[, 1]),
               end = as.numeric(str_split_fixed(coord, "-", 2)[,2]) ) %>%
        mutate(QTL_junction = paste0(QTL_junction, ":", coord))
    }
    stopifnot(nrow(coord_df) > 0)
    # if SUPPA - swap back junctions_loc with new coord_df 
    coord_df$GWAS_SNP_Beta <- junctions_loc$GWAS_SNP_Beta[ match(coord_df$feature, junctions_loc$feature)] 
    junctions_loc <- coord_df
    # HERE - figure out orientation of AF/AL/A5/A3 to get directionality the right way around.
    
    stopifnot(nrow(junctions_loc) > 0)
    
    ## Plot a bar
    grange_loc <-  
      GRanges(seqnames = junctions_loc$chr, 
              ranges = IRanges(start = junctions_loc$start, end = junctions_loc$end), 
              #PP.H4.abf = junctions_loc$PP.H4.abf, 
              #SNP = junctions_loc$QTL_SNP, 
              junction_id = junctions_loc$QTL_junction, 
              QTL_Beta = junctions_loc$GWAS_SNP_Beta)
    
    
    grange_loc$junction_id <- factor(grange_loc$junction_id)#, levels = rev(junctions_loc$QTL_junction))
    plot <- ggplot() +
      scale_colour_distiller(name = "QTL Beta", limits = mylim, palette = "RdBu") +
      scale_fill_distiller(name = "QTL Beta", limits = mylim, palette = "RdBu") +
      
      theme_classic() + 
      theme(
        axis.text = element_text(colour = "black"), 
        legend.key.size = unit(0.25, 'cm'), 
        legend.title = element_text(size=10) 
      )  + 
      geom_alignment(data = grange_loc, aes_string(group = "junction_id", fill = colourby, colour = colourby),
                     range.geom = "rect", width = 5 )
    
    if(guide == FALSE){
      plot <- plot + guides(colour = "none", fill = "none")
    }
    
    return(plot)
  }

#junction_plot(mylocus = "PLCG2", type = "A5", flank = 10)
#junction_plot(mylocus = "SIGLEC11", type = "A5", flank = 10)

junction_plot(mylocus = "CTSH", type = "SE", flank = 1)
# tracks(
# "junctions" = junction_plot(mylocus = "PLCG2", type = "leafcutter", gwas = "Bellenguez_2021"),
# "af" = junction_plot(mylocus = "PLCG2", type = "AF", gwas = "Bellenguez_2021"),
# "se" = junction_plot(mylocus = "PLCG2", type = "SE", gwas = "Bellenguez_2021")
# )


snp_plot <- function(mylocus, snp_width = 50, type = "GWAS", n = Inf, PP4 = 0.5){
  if(type == "GWAS"){
  snp_loc <- filter(all_coloc, locus  == mylocus) %>%
    dplyr::select(SNP = GWAS_SNP, chr = GWAS_chr, end = GWAS_pos, GWAS) %>%
    distinct() %>%
    mutate(label = paste0(GWAS, "\n", SNP) )
    #tidyr::separate(QTL_junction, into = c("chr", "start", "end"), sep = ":|-", remove = FALSE ) %>%
    #mutate(start = as.numeric(start), end = as.numeric(end)) %>%
    #arrange(desc(PP.H4.abf)) %>%
    #head(n)
  stopifnot(nrow(snp_loc) > 0)
  
  ## Plot a bar
  grange_loc <-  
    GRanges(seqnames = snp_loc$chr, 
            ranges = IRanges(start = snp_loc$end - snp_width, end = snp_loc$end), 
            SNP = snp_loc$SNP, GWAS = snp_loc$GWAS, label = snp_loc$label )
  }
  if(type == "QTL"){
  snp_loc <- filter(all_coloc, locus  == mylocus, PP.H4.abf > PP4) %>%
      dplyr::select(SNP = QTL_SNP, chr = QTL_chr, end = QTL_pos, PP.H4.abf) %>%
      distinct() %>%
    arrange(desc(PP.H4.abf)) %>%
    head(n)  %>%
    mutate(label = SNP )
    #tidyr::separate(QTL_junction, into = c("chr", "start", "end"), sep = ":|-", remove = FALSE ) %>%
    #mutate(start = as.numeric(start), end = as.numeric(end)) %>%
    #arrange(desc(PP.H4.abf)) %>%
    #head(n)
    stopifnot(nrow(snp_loc) > 0)
    
    ## Plot a bar
    grange_loc <-  
      GRanges(seqnames = snp_loc$chr, 
              ranges = IRanges(start = snp_loc$end - snp_width, end = snp_loc$end), 
              SNP = snp_loc$SNP, label = snp_loc$label )
    
  }
  
  #grange_loc$junction_id <- factor(grange_loc$junction_id, levels = rev(junctions_loc$QTL_junction))
  grange_loc$fillby <- 0
  plot <- 
    ggplot() + 
    geom_alignment(data = grange_loc, aes(group = label, fill = fillby), range.geom = "rect" ) +
    theme_classic() + scale_fill_continuous() +
    theme(legend.text = element_blank(), axis.text = element_text(colour = "black") ) + 
    labs(fill = "")# +
    #guides(colour = "none", fill = "none")
    #scale_fill_viridis_c() + theme_bw() #+# theme(legend.position = "top")

  return(plot)
}

snp_plot(mylocus = "PLCG2",type = "QTL", PP4 = 0.9)


beta_plot <- function(mylocus, PP4 = 0.8, mylim = c(-1,1), myset = NULL, myref = NULL){
  
  to_plot <- all_coloc %>%
  filter(locus == mylocus, PP.H4.abf > PP4) %>%
    mutate(feature = gsub("\\.[0-9]+:", "\n",feature)) %>%
    mutate(feature = paste0(gene_name, "\n", feature)) 
  if(!is.null(myset)){
    to_plot <- filter(to_plot, set %in% myset)
  }
  if(!is.null(myref)){
    to_plot <- filter(to_plot, reference %in% myref) %>%
      mutate(reference = factor(reference, levels = myref))
  }
  
  to_plot %>%
  ggplot(aes(x = GWAS_SNP_Beta, y = feature)) + 
  geom_point() + 
  geom_errorbarh(aes(xmin = GWAS_SNP_Beta - GWAS_SNP_SE, xmax = GWAS_SNP_Beta + GWAS_SNP_SE), height = 0.5) +
  scale_x_continuous(limits = mylim) +
  geom_vline(xintercept = 0, linetype = 3) + theme_classic() +
  facet_grid(set ~ reference, scales = "free", space = "free")

}
beta_plot("PLCG2", PP4 = 0.5)

```



```{r}


isoform_plot(mylocus = "SIGLEC11", transcript_type = "isoseq", level = "transcript", PP4 = 0.8, mylim = c(-1.2,1.2) )#, colourby = "transcript_type" )

isoform_plot(mylocus = "CD33", transcript_type = "isoseq", level = "transcript", PP4 = 0.9, mylim = c(-1.2,1.2), gwas = "Lambert_2013")

isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, guide = TRUE)

isoform_plot(mylocus = "PLCG2", mygene = c("PLCG2"), transcript_type = "isoseq", level = "transcript", PP4 = 0.8)

isoform_plot(mylocus = "PLCG2", transcript_type = "isoseq", level = "transcript", PP4 = 0.8)
isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8)

```

# PLCG2

```{r}

beta_plot("PLCG2")
finemap_plot("PLCG2-2")


plcg2_beta <- beta_plot("PLCG2", PP4 = 0.5, myset = c("expression", "transcript", "leafcutter"), myref = "GENCODE+Novel", mylim = c(-0.6,0.6)) + theme_jh()
ggsave(plot = plcg2_beta, filename = "plots/PLCG2_beta_plot.pdf", width = 80, height = 40, units = "mm")


#gene_meta
plc <- filter(all_coloc, locus == "PLCG2", PP.H4.abf > 0)
# tx <- union_gff[ union_gff$gene_id %in% c("ENSG00000197943.10", "ENSG00000261218.5") & union_gff$type == "exon" & union_gff$transcript_id %in% all_coloc$feature]
# 
# tx$GWAS_Beta <- plc$GWAS_SNP_Beta[ match(tx$transcript_id, plc$feature)]
# tx$GWAS_SE <- plc$GWAS_SNP_SE[ match(tx$transcript_id, plc$feature)]
# tx$PP4 <- plc$PP.H4.abf[ match(tx$transcript_id, plc$feature)]
# 
# library(patchwork)
# 
# tx_plot <- tx %>%
#       ggplot(aes(
#         xstart = start,
#         xend = end,
#         y = transcript_name
#     )) +
#     geom_range(
#         aes(colour = GWAS_Beta)
#     ) +
#     geom_intron(arrow.min.intron.length = 5000,
#         data = to_intron(as.data.frame(tx), "transcript_name"),
#         aes(strand = strand)
#     ) +
#   theme_classic() 
# 
# tx_stats <- 
#   ggplot(tx, aes(y = transcript_id, x = GWAS_Beta)) + 
#   geom_errorbarh(aes(xmin = GWAS_Beta - GWAS_SE, xmax = GWAS_Beta + GWAS_SE)) +
#   geom_point(aes(colour = PP4, shape = PP4 > 0.5), size = 5) +
#   theme_classic() +
#   geom_vline(xintercept = 0, linetype = 3)

isoform_plot(mylocus = "PLCG2", transcript_type = "isoseq", level = "expression", PP4 = 0.9)
isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.9)

isoform_plot(mylocus = "PLCG2", transcript_type = "isoseq", level = "transcript", PP4 = 0.9)

plcg2_plot <- function(x_lim, betalim){
  tracks(
   # eQTLs = isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "expression", PP4 = 0.8),
    "tuQTLs" = isoform_plot(mylocus = "PLCG2", transcript_type = "isoseq", level = "transcript", PP4 = 0.8),
    #"AF" = junction_plot(mylocus = "PLCG2", type = "AF", gwas = "Bellenguez_2021"),
    "leafcutter" = lc_plot("PLCG2"), 
    #"A5" = junction_plot(mylocus = "PLCG2", type = "A5", gwas = "Bellenguez_2021", flank = 1),
   #"SE" = junction_plot(mylocus = "PLCG2", type = "SE", gwas = "Bellenguez_2021"),
    #"leafcutter" = junction_plot("PLCG2", "Bellenguez_2021", type = "leafcutter"),
    "QTL SNPs" = snp_plot(mylocus = "PLCG2",type = "QTL", snp_width = 1, PP4 = 0.8 ),
    #GWAS =  snp_plot(mygene = "PLCG2", mylocus = "PLCG2",type = "GWAS", snp_width = 1),
    "GWAS SNPs" = finemap_plot("PLCG2-2", 10),
    #"caQTLs" = peak_plot("PLCG2", PP4 = 0.5 ),
    #xlim = c(81739000,81959116),# full gene
    #xlim = c(81739000,81785000), # first section
   #xlim = c(81739000,81740000), # close up
   xlim = x_lim,
    heights = c(3,3,1,1), 
   title = "PLCG2"
  ) + theme(plot.title = element_text(face = "italic"))
}

plcg2_overview_plot <-   tracks(
 Genes = isoform_plot(mylocus = "PLCG2", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = c(0,0)),
# "Transcripts" = isoform_plot(mylocus = "PLCG2", transcript_type = "isoseq", level = "transcript", PP4 = 0.8 ),
  "GWAS SNPs" = finemap_plot("PLCG2-2", 10),
   #   "caQTLs" = peak_plot("PLCG2", PP4 = 0.5 ),
   xlim = c(81739000,81959116)
)
plcg2_overview_plot 

#plcg2_plot( c(81739000,81959116) )  #whole gene
p1 <- plcg2_plot( c(81739000,81785000) ) # intron 1 - 3
p2 <- plcg2_plot( c(81739200,81740000) ) + theme(axis.text.y = element_blank() ) # intron 1 only

ggsave(plot = p1, filename = here::here("plots/PLCG2_intron1_3_plot.pdf"), height = 4,width = 6.5, dpi = 600)

ggsave(plot = p2, filename = here::here("plots/PLCG2_exon1_plot.pdf"), height = 7,width = 5, dpi = 600)

ggsave(plot = plcg2_overview_plot,  filename = here::here("plots/PLCG2_full_gene.pdf"), height = 3,width = 10)
p1
p2
```

Bellenguez found 2 independent PLCG2 loci:
rs72824905 - rare missense variant in middle of gene (MAF = 0.008; OR = 0.74)
rs12446759 - common intergenic variant outside of gene body (MAF = 0.43; OR = 0.95)


# CD33

```{r}

filter(all_coloc, locus == "CD33", PP.H4.abf > 0.8) %>%
  select(set, reference, feature, GWAS_SNP_Beta, PP.H4.abf)


beta_plot("CD33",mylim = c(-1.2,1.2))
cd33_lim <- c(-1.2,1.2)

cd33_plot <- function(x_lim, betalim){
  p <- tracks(
  #Genes = isoform_plot(mylocus = "CD33", transcript_type = "expression", level = "gene", PP4 = 0.6),
  "transcript QTLs" = isoform_plot(mylocus = "CD33", transcript_type = "isoseq", level = "transcript", PP4 = 0.5, mylim = cd33_lim, gwas = "Lambert_2013"),
 # "GENCODE\ntranscripts" = isoform_plot(mylocus = "CD33", mygene = "CD33", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8, mylim = cd33_lim, gwas = "Lambert_2013"),
  "junction QTLs" = lc_plot("CD33", "Lambert_2013", type = "leafcutter",mylim = cd33_lim),
  "Exon skipping" = junction_plot("CD33", type = "SE", gwas = "Lambert_2013",mylim = cd33_lim),
  "Intron Retention" = junction_plot("CD33", type = "RI", gwas = "Lambert_2013",mylim = cd33_lim),
  #A3 = junction_plot("CD33", type = "A3", gwas = "Lambert_2013",mylim = cd33_lim),
  #A5 = junction_plot("CD33", type = "A5", gwas = "Lambert_2013",mylim = cd33_lim),
  "QTL SNPs" = snp_plot( mylocus = "CD33",type = "QTL", snp_width = 1, PP4 = 0.8),
  "GWAS SNPs" =  snp_plot(mylocus = "CD33",type = "GWAS", snp_width = 1),
  #xlim = c(51211000,51240000),
  #xlim = c(51222500, 51227500), # close up,
  xlim = x_lim,
  heights = c(1.5,1.5,0.5,0.5,1.5,0.75)
)
return(p)
}  

cd33_far <- cd33_plot( c(51224500,51240000) )

cd33_close <- cd33_plot( c(51224500, 51226600))
cd33_beta <- beta_plot("CD33",mylim = c(-1.2,1.2))

ggsave(plot = cd33_beta, filename = "plots/cd33_coloc_beta.pdf", width = 5, height = 5)

ggsave(plot = cd33_close, filename = "plots/cd33_coloc_close_up.pdf", width = 8, height = 5)
ggsave(plot = cd33_far, filename = "plots/cd33_coloc_far.pdf", width = 8, height = 5)

```

Plot the raw SUPPA PSI values

```{r}
suppa_ri <- read.table("data/suppa/roussos_suppa_ri.psi", header=TRUE)
suppa_ri$ID <- gsub(";", ":", row.names(suppa_ri))

suppa_se <- read.table("data/suppa/roussos_suppa_se.psi", header=TRUE)
suppa_se$ID <- gsub(";", ":", row.names(suppa_se))

cd33_vcf <- read.table("data/vcf/roussos_v2_unfiltered_cd33.vcf", header=TRUE) %>%
  select(-INFO, -FORMAT, -FILTER, -QUAL, -CHROM, -POS) %>%
  pivot_longer(names_to = "donor", values_to = "value", cols = !c(ID,REF,ALT) ) %>%
  mutate(genotype = case_when(
    grepl("0\\|0", value) ~ paste0(REF, "/",REF),
    grepl("0\\|1|1\\|0", value) ~ paste0(REF, "/",ALT),
    grepl("1|1", value) ~ paste0(ALT, "/",ALT)
  ) ) %>%
  filter(donor %in% names(suppa_ri), ID %in% c("rs3865444")) %>%
mutate(genotype = factor(genotype, 
                            levels = c(
                              paste0(unique(REF),"/", unique(REF) ), 
                              paste0(unique(REF), "/", unique(ALT) ), 
                              paste0(unique(ALT),"/",unique(ALT) )
                              ) 
                            )
)
       



cd33_coloc <- filter(all_coloc, locus == "CD33", PP.H4.abf > 0.8) %>%
  select(set, reference, feature, GWAS_SNP_Beta, PP.H4.abf)

cd33_psi <- filter(bind_rows(suppa_ri, suppa_se), ID %in% cd33_coloc$feature) %>%
  pivot_longer(cols = !(ID), names_to = "donor", values_to = "psi") %>%
  pivot_wider(names_from = "ID", values_from = "psi") %>%
  left_join(cd33_vcf, by = "donor")

library(patchwork)
cd33_exon <- 
cd33_psi %>%
  ggplot(aes(x = genotype, y = `ENSG00000105383.15:SE:chr19:51225155-51225218:51225598-51225803:+`)) + 
  geom_violin(scale = "width", fill = "skyblue") +
  geom_jitter(width = 0.1, height = 0) +
  geom_boxplot(width = 0.25, fill =NA, colour = "white",outlier.colour = NA) +
  labs(y = "% Exon 2 inclusion", x = "rs3865444" ) +
  theme_classic()

cd33_intron <- 
cd33_psi %>%
  ggplot(aes(x = genotype, y = `ENSG00000105383.15:RI:chr19:51225098:51225155-51225218:51225598:+`)) + 
   geom_violin(scale = "width", fill = "firebrick") +
  geom_jitter(width = 0.1, height = 0) +
  geom_boxplot(width = 0.25, fill =NA, colour = "white", outlier.colour = NA) +
  labs(y = "% Intron retention", x = "rs3865444") +
  theme_classic()

cd33_compare <- 
cd33_psi %>% 
  ggplot(aes(x = `ENSG00000105383.15:RI:chr19:51225098:51225155-51225218:51225598:+`, 
             y = `ENSG00000105383.15:SE:chr19:51225155-51225218:51225598-51225803:+`, 
             colour = genotype, group = genotype )) + 
  geom_point() + 
  labs(x = "% Intron retention", y = "% Exon 2 inclusion") + 
  theme_classic() +
  scale_colour_viridis_d() +
  scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
  geom_smooth(method = "lm",se = FALSE) +
  ggpubr::stat_cor(label.y.npc = 0.1)

cd33_combine_plot <- cd33_exon + cd33_intron + cd33_compare + 
  plot_layout(nrow = 1) &
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) &
  theme(axis.text = element_text(colour = "black"))

ggsave(plot = cd33_combine_plot, filename = "plots/cd33_coloc_combine.pdf", width = 10, height = 3)

```




Already established that the causal SNP rs12459419 sits close to 3' splice site of exon 2 and increases exon 2 skipping. However, long reads show that in addition to exon 2 skipping, there is an extension of exon 1 via intron retention, presumably due to a weakening of the e1-e2 3' splice site. 
Transcripts with exon1-2 intron retention have a predicted start codon downstream of the canonical transcript, thus they lose ~16 amino acids at the N terminal.   
Uniprot suggests that the first 17 amino acids encodes a signalling peptide, so the variant would cause a shift into production of protein without the peptide.
Intron retention previously observed by Malik et al HMG 2015 https://academic.oup.com/hmg/article/24/12/3557/622473
but they assumed that the IR would lead to NMD, but in the same paper they show that it doesn't undergo NMD.

What is role of SIGLEC22P sequence?
SIGLEC22P activation by CRISPR http://doi.org/10.1089/crispr.2021.0052

How to prove the truncated peptide is produced? Ribosome footprinting data might show engagement at the downstream start codon.  

July 2022 updates - reran QTLs using hybrid reference with GENCODE plus all novel transcripts from the isoseq reference, excluding the fusion transcripts due to their overestimation both by Bambu and by Kallisto.
The result for CD33 is that because all the intron retention transcripts were part of the fusion, there is no more intron retention to be tested for QTLs. Unclear what to do.

If I bring back the fusions, then people may claim they're artifactual.


# MS4A6A


```{r}
filter(all_coloc, locus == "MS4A6A", PP.H4.abf > 0.6) %>%
  select(set, feature, GWAS_SNP_Beta, PP.H4.abf)

finemap_plot("MS4A4A")
beta_plot("MS4A6A", PP4 = 0.5)

ms4a6a_plot <- function(x_lim, beta_lim, pp4 = 0.8){
tracks(
  #Genes = isoform_plot(mylocus = "MS4A6A", transcript_type = "isoseq", level = "expression", PP4 = 0.6),
#  "isoseq\ntranscripts" = isoform_plot(mylocus = "MS4A6A", mygene = "MS4A6A", transcript_type = "isoseq", level = "transcript", PP4 = pp4, mylim = beta_lim),
  #"GENCODE\ntranscripts" = isoform_plot(mylocus = "MS4A6A", mygene = "MS4A6A", transcript_type = "GENCODE_transcript", level = "transcript", PP4 = 0.8),
  #A5 = junction_plot("MS4A6A", type = "A5", PP4 = pp4, mylim = beta_lim ),
  #RI = junction_plot("MS4A6A", type = "RI", PP4 = pp4, mylim = beta_lim ),
  junctions = junction_plot("MS4A6A",type = "leafcutter", PP4 = pp4, mylim = beta_lim),
  QTL = snp_plot(mylocus = "MS4A6A",type = "QTL", snp_width = 1, PP4 = pp4),
  GWAS =  snp_plot(mylocus = "MS4A6A",type = "GWAS", snp_width = 1),
  "SuSiE" = finemap_plot("MS4A4A", label = TRUE, min_pip = 0),
  xlim = x_lim,
  #heights = c(1,1,0.5,1,1.5,1,1.5),
  title = "MS4A6A - PP4 > 0.5"
 # xlim = c(60160000, 60320000),
  #xlim = c(51222500, 51227500),
# heights = c(2,1,1,0.5,1,1.5, 1)
)
}

ms4a6a_plot(c(60169607,60186666), pp4 = 0.5, beta_lim = c(-0.3,0.3) )
ms4a6a_plot(c(60172947,60178376), pp4 = 0.5, beta_lim = c(-0.3,0.3) )

```

## CTSH

```{r}
beta_plot("CTSH", PP4 =  0.8)
finemap_plot("CTSH", col_width = 500)

ctsh <- filter(all_coloc, PP.H4.abf > 0.8, locus == "CTSH")

ctsh_plot <- function(x_lim,beta_lim = c(-0.5,0.5), label_snps = FALSE){
tracks(
  "tuQTL" =  isoform_plot(mylocus = "CTSH", mygene = "CTSH", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8, mylim = beta_lim),
  SE = junction_plot("CTSH", type = "SE", PP4 = 0.8, mylim = beta_lim, flank = 10 ),
  AF = junction_plot("CTSH", type = "AF", PP4 = 0.8, mylim = beta_lim, flank = 10  ),
  A3 = junction_plot("CTSH", type = "A3", PP4 = 0.8 , mylim = beta_lim, flank = 10),
  #RI = junction_plot("CTSH", type = "RI", PP4 = 0.8 ),
  leafcutter = lc_plot("CTSH",type = "leafcutter", PP4 = 0.8, mylim = beta_lim),
  "QTLs" = snp_plot(mylocus = "CTSH",type = "QTL", snp_width = 1, PP4 = 0.8),
  "GWAS" = finemap_plot("CTSH", label = label_snps, xlim = x_lim, min_pip = 0.1),
  xlim = x_lim, # highlight A3
  heights = c(2.5,0.75,1,0.5,1.2,1,1)
)
}

#ctsh_plot(c(78921000, 78949500))
ctsh_plot(c(78921000, 78949500)) # full
ctsh_plot(c(78936500, 78945500), label_snps = TRUE) # close up 
ctsh_plot(c(78939000, 78945500)) # highlight A3

ctsh_plot(c(78943819,78945643), label_snps = TRUE )


isoform_plot("CTSH", transcript_type = "GENCODE", PP4 = 0, level = "transcript",  mygene = "CTSH", colourby = "status")

```

Jul 2022 update - intron retention is now missing?

Multiple QTLs colocalise here - some difficult to reconcile together. 
Leafcutter junction usage points to an alternative 5' splice site of the first 
lead leafcutter SNP is rs34593439
lead  transcript usage and SUPPA QTL SNP is rs2289702 which sits 60bp upstream of the splice site.
GWAS SNP rs12592898 associated with increased use of short first exon.
all 3 SNPs are in high LD (R2 = 0.78-0.98, European populations in LDlink).

Use of longer first exon shifts start codon downstream, so splicing would change amino acid composition by 30+ residues.
According to Uniprot the first 22 residues encode a signalling peptide, so could have functional consequences.


## PILRA

```{r}
beta_plot("PILRA", mylim = c(-2,2))
finemap_plot("NYAP1")

pilra_plot <- function(x_lim){
  tracks(
  "eQTL" = isoform_plot(mylocus = "PILRA", transcript_type = "isoseq", level = "expression", PP4 = 0.8, mylim = c(-1,2)),
  "tuQTL" = isoform_plot(mylocus = "PILRA", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8, mylim = c(-1,2)),
  "QTL SNP" = snp_plot(mylocus = "PILRA",type = "QTL", snp_width = 1, PP4 = 0.8),
  "SUSiE" = finemap_plot("NYAP1", col_width = 1, label = TRUE),
  xlim = x_lim
  )
}

pilra_plot(c(100373311,100374707))

pilra_plot(c(100318900,100401484))

```


## NME8/GPR141/EPDR1

```{r}
beta_plot("EPDR1")

gpr141_plot <- function(x_lim, beta_lim = c(-0.4,0.4)){
tracks(
  eQTL = isoform_plot(mylocus = "EPDR1", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  tuQTL = isoform_plot(mylocus = "EPDR1", transcript_type = "isoseq", level = "transcript", PP4 = 0.8, mylim = beta_lim),
  A3 = junction_plot("EPDR1", type = "A3", PP4 = 0.8 ),
  "QTL SNP" = snp_plot(mygene = "GPR141", mylocus = "EPDR1",type = "QTL", snp_width = 1, PP4 = 0.8),
  finemap = finemap_plot("NME8"),
  xlim = x_lim
  #xlim = c(37659249,37864889),
  #xlim = c(37683500,37685800) # close up 
)
}

gpr141_plot( c(37659249,37864889) )
gpr141_plot( c(37684500,37685800) )
```

All roads point to GPR141, a relatively unknown gene, being causal here.
It's an eQTL 

## SIGLEC11

```{r}
beta_plot("SIGLEC11", PP4 = 0.5)




siglec11_plot <- function(x_lim, beta_lim = c(-0.6,0.6)){
  tracks(
  eQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  tuQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "isoseq", level = "transcript", PP4 = 0.8, mylim = beta_lim),
  A3 = junction_plot("SIGLEC11", type = "A5", PP4 = 0.8, flank = 10 ),
  "QTL SNPs" = snp_plot( mylocus = "SIGLEC11",type = "QTL", snp_width = 1, PP4 = 0.8),
  "GWAS SNPs" = finemap_plot("SIGLEC11", label = TRUE, min_pip = 0.01, xlim = x_lim, col_width = 10),
  xlim = x_lim
)
}

siglec11_plot(x_lim = c(49946825,49961113)) 
siglec11_plot(x_lim = c(49958117,49961236)) 


siglec11 <- filter(all_coloc, locus == "SIGLEC11", PP.H4.abf > 0)
tx <- union_gff[ union_gff$gene_id %in% c("ENSG00000161640.15") & union_gff$type %in% c("exon", "CDS") & union_gff$transcript_id %in% all_coloc$feature]

tx$GWAS_Beta <- siglec11$GWAS_SNP_Beta[ match(tx$transcript_id, siglec11$feature)]
tx$GWAS_SE <- siglec11$GWAS_SNP_SE[ match(tx$transcript_id, siglec11$feature)]
tx$PP4 <- siglec11$PP.H4.abf[ match(tx$transcript_id, siglec11$feature)]

cds <- as.data.frame(tx[ tx$type == "CDS"])
exons <- as.data.frame(tx[tx$type == "exon"])

exons %>% 
      ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_id
    )) +
    geom_intron(arrow.min.intron.length = 5000,
        data = to_intron(as.data.frame(tx), "transcript_name"),
        aes(strand = strand)
    ) +
    geom_range(height = 0.25,
        aes(fill = GWAS_Beta)
    ) +
    geom_range(data = cds,
                aes(fill = GWAS_Beta)
               ) +

  theme_classic() 

tx_stats <- 
  ggplot(tx, aes(y = transcript_id, x = GWAS_Beta)) + 
  geom_errorbarh(aes(xmin = GWAS_Beta - GWAS_SE, xmax = GWAS_Beta + GWAS_SE)) +
  geom_point(aes(colour = PP4, shape = PP4 > 0.5), size = 5) +
  theme_classic() +
  geom_vline(xintercept = 0, linetype = 3)

tx_plot 
tx_stats

```

fine-mapped SNPs overlapping or close to exons :
rs9304690 - synoymous variant in final exon / 3'UTR
rs62113133 - nonsynymous variant in exon 6.

## TREM2

```{r}
beta_plot("TREM2", PP4 = 0.5, mylim = c(-2,2) )


trem2_coloc <- filter(all_coloc, locus == "TREM2", PP.H4.abf > 0.5)

trem2_plot <- function(x_lim){
tracks(
  isoseq_transcripts = isoform_plot(mylocus = "TREM2", PP4 = 0, mylim = x_lim, transcript_type = "isoseq", level = "transcript"),
  leafcutter = junction_plot("TREM2",type = "leafcutter", PP4 = 0.8),
  "QTL SNP" = snp_plot(mylocus = "TREM2",type = "QTL", snp_width = 1, PP4 = 0.8),
  "SuSiE" = finemap_plot("TREM2-1", label = TRUE, xlim = x_lim,col_width = 5 ),
  xlim = x_lim,
  heights = c(4,1,1,1)
)
  }

trem2_plot(c(41148253,41163428))
#trem2_plot(c(41148253,41190282))
trem2_plot( c(41161207,41161689) )

trem2 <- filter(sqanti, grepl("ENSG00000095970.17", associated_gene))

```

Variant	Chromosome	Position	Gene locus	Minor/major allele	MAF	OR	95% CI	P value
rs10947943	6	41036354	UNC5CL	TREM2	A/G	0.142	0.95 (0.93-0.97)	6.21E-06
rs143332484	6	41161469	TREM2	TREM2	T/C	0.013	1.41	1.32–1.50	2.8 × 10−25
rs75932628	6	41161514	TREM2	TREM2	T/C	0.003	2.39	2.09–2.73	2.5 × 10−37
rs60755019	6	41181270	TREML2	TREM2	G/A	0.004	1.55	1.33–1.80	2.1 × 10−8

rs75932628 is the R47H coding variant with MAF = 0.003 - too rare to be picked up in QTL work
However, rs143332484 is an order of magnitude more common - MAF = 0.013 and is also within TREM2 exon 2.
The final SNP is rs60755019, rare and far outside the coding sequence

Best sQTL is the exon 2 skipping intron, with lead SNP rs188904277 in high LD with rs143332484

Fine-mapping gives the highest PIP to  chr6:41161469:C:T, a SNP without an rsid in my build, but manually looking for it, it's labelled as rs143332484 . It's a known missense mutation R62H. It's also in high LD (0.82) with lead sQTL SNP rs188904277

TEST ALL TREM2 transcript TPMs against chr6:41161469:C:T

```{r}

trem2_tx <- read_tsv("data/TREM2/roussos_trem2_transcript_tpm.tsv")
roussos_sk <- read_tsv("data/TREM2/roussos_185_mbv_no_dups_sample_key.tsv")

trem2_vcf <- read.table("data/TREM2/roussos_v2_unfiltered_trem2_genotypes.vcf", header= TRUE) %>%
    select(-INFO, -FORMAT, -FILTER, -QUAL, -CHROM, -POS) %>%
  pivot_longer(names_to = "donor", values_to = "value", cols = !c(ID,REF,ALT) ) %>%
  mutate(genotype = case_when(
    grepl("0\\|0", value) ~ paste0(REF, "/",REF),
    grepl("0\\|1|1\\|0", value) ~ paste0(REF, "/",ALT),
    grepl("1|1", value) ~ paste0(ALT, "/",ALT)
   ) )  %>%
  filter(ID == c("chr6:41161469:C:T;rs143332484"), donor %in% roussos_sk$participant_id) %>%
  mutate(genotype = factor(genotype,
                            levels = c(
                              paste0(unique(REF),"/", unique(REF) ),
                              paste0(unique(REF), "/", unique(ALT) ),
                              paste0(unique(ALT),"/",unique(ALT) )
                              )
                            )
  )

# individual TPMs
trem2_tx[c("transcript_id", roussos_sk$sample_id)] %>%
  pivot_longer(names_to = "sample_id", values_to = "TPM", cols = !c(transcript_id)) %>%
  left_join(roussos_sk, by = "sample_id") %>%
  left_join(trem2_vcf, by = c("participant_id" = "donor")) %>%
  ggplot(aes(x = genotype, y = TPM)) + geom_jitter(width = 0.2) + geom_boxplot(fill = NA, outlier.colour = NA) +
facet_wrap(~transcript_id, scales = "free") 
  
# relative proportions
trem2_tu <- sweep(trem2_tx[,-1], MARGIN = 2, STATS = colSums(trem2_tx[,-1]), FUN = "/")
trem2_tu$transcript_id <- trem2_tx$transcript_id

trem2_tu[c("transcript_id", roussos_sk$sample_id)] %>%
  pivot_longer(names_to = "sample_id", values_to = "TPM", cols = !c(transcript_id)) %>%
  left_join(roussos_sk, by = "sample_id") %>%
  left_join(trem2_vcf, by = c("participant_id" = "donor")) %>%
  ggplot(aes(x = genotype, y = log2(TPM + 0.1) )) + geom_jitter(width = 0.2) + geom_boxplot(fill = NA, outlier.colour = NA) +
facet_wrap(~transcript_id, scales = "free") 
  


```




# PRKD3

```{r}
beta_plot("PRKD3", mylim = c(-5,5))
```


## Parkinson's 

### BST1
```{r}
beta_plot("BST1", mylim = c(-2,2))
```

no GWAS beta/se here

## FAM49B/CYRIB

```{r}
beta_plot("FAM49B")

fam_finemapping <- read_csv(here::here("external/Schilder_PD_finemapping/FAM49B.Nalls23andMe_2019.UKB.multi_finemap.csv") )
fam_hg38_snps <- read_tsv(here::here("external/Schilder_PD_finemapping/FAM49B.Nalls23andMe_2019.UKB.multi_finemap.hg38.bed"), col_names = c("chr", "start", "end", "SNP"))

fam_df <- left_join(fam_finemapping, fam_hg38_snps, by = "SNP")



fam_gwas_plot <- fam_df %>%
    ggplot(aes(x = end , y = -log10(P) )) + 
  geom_point(aes(colour = r2), size = 0.5 ) +
  geom_point(data = filter(fam_df, leadSNP == TRUE),  colour = "black", fill = "purple", shape = 23, size = 2) +
  theme_classic() +
  scale_colour_distiller(palette = "RdBu") + 
  labs(colour = "LD R2", x = "position", y = expression(log[10]~P) )

fam_finemap_plot <- 
  fam_df %>%
    ggplot(aes(x = end , y = mean.PP )) + 
  geom_point(aes(colour = 1), size = 0.5 ) +
  geom_point(data = filter(fam_df, leadSNP == TRUE),  colour = "black", fill = "purple", shape = 23, size = 2) +
  theme_classic() +
  labs(colour = "PIP", x = "position", y = "PIP" )

fam_gwas_plot + 
fam_finemap_plot +
  plot_layout(ncol = 1)


fam_plot <- function(x_lim, beta_lim = c(-0.3,0.3)){
tracks(
  #eQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  tuQTL = isoform_plot(mylocus = "FAM49B", transcript_type = "GENCODE", level = "transcript", PP4 = 0.9, gwas = "Nalls23andMe_2019"),
  #tuQTL = isoform_plot(mylocus = "FAM49B", transcript_type = "isoseq", level = "transcript", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  leafcutter = lc_plot("FAM49B", type = "leafcutter", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  "QTL SNP" = snp_plot( mylocus = "FAM49B",type = "QTL", snp_width = 1, PP4 = 0.8),
  SE = junction_plot("FAM49B", type = "SE", PP4 = 0.8,gwas = "Nalls23andMe_2019", flank  =1 ),
  AF = junction_plot("FAM49B", type = "AF", PP4 = 0.8,gwas = "Nalls23andMe_2019", flank  =1 ),
 #   "GWAS SNP" = snp_plot( mylocus = "FAM49B",type = "GWAS", snp_width = 1, PP4 = 0.8),
 "GWAS" = fam_gwas_plot, 
 "finemap" = fam_finemap_plot,
  xlim = x_lim
  )
}

fam_plot(c(129831146,130021477))
fam_plot(c(129900000,129910000))
  #finemap = finemap_plot("SIGLEC11", label = TRUE, min_pip = 0.01, xlim = x_lim),
#  xlim = x_lim
#)
```

## SIPA1L2

MAIN FIGURE

```{r}
sipa_finemap <- read_csv(here::here("external/Schilder_PD_finemapping/SIPA1L2.Nalls23andMe_2019.UKB.multi_finemap.csv") )
sipa_coords <- read_tsv(here::here("external/Schilder_PD_finemapping/SIPA1L2.Nalls23andMe_2019.UKB.multi_finemap.hg38.bed"), col_names = c("chr", "start", "end", "SNP", "X") )


# put in hg38 coordinates
sipa_finemap <- left_join(sipa_finemap, sipa_coords, by = "SNP")

sipa_gwas_df<- sipa_finemap %>%
  mutate(chr = paste0("chr", CHR), start, end, score = -log10(P) ) %>%
  select(chr, start, end, score, SNP)

write_tsv( select(sipa_gwas_df, -score), file = "external/Schilder_PD_finemapping/SIPA1L2.gwas.bed", col_names = FALSE)
write_tsv( select(sipa_gwas_df, -SNP), file = "external/Schilder_PD_finemapping/SIPA1L2.gwas.bedGraph", col_names = FALSE)



sipa_gwas_plot <- sipa_finemap %>%
  ggplot(aes(x = end , y = -log10(P) )) + 
  geom_point(aes(colour = r2), size = 0.5 ) +
  geom_point(data = filter(sipa_finemap, leadSNP == TRUE),  colour = "black", fill = "purple", shape = 23, size = 2) +
  theme_classic() +
  scale_colour_distiller(palette = "RdBu") + 
  labs(colour = "LD R2", x = "position", y = expression(log[10]~P) )

sipa_finemap_plot <-
  sipa_finemap %>%
  ggplot(aes(x = end , y = SUSIE.PP )) + 
  geom_point(aes(colour = 1), size = 1 ) +
  labs(x = "position", y = "PIP") +
  theme_bw() 

sipa_gwas_plot + sipa_finemap_plot +
  plot_layout(ncol = 1)

#  geom_text_repel(data = filter(sipa_finemap, Support > 0), aes(label = SNP))
isoform_plot(mylocus = "SIPA1L2", mygene = "SIPA1L2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.1, gwas = "Nalls23andMe_2019")
junction_plot("SIPA1L2", type = "SE", PP4 = 0.9, gwas = "Nalls23andMe_2019")
junction_plot("SIPA1L2", type = "SE", PP4 = 0.9, gwas = "Nalls23andMe_2019")
junction_plot("SIPA1L2", type = "A5", PP4 = 0.8, gwas = "Nalls23andMe_2019", flank = 1 )
beta_plot("SIPA1L2",PP4 = 0.8)

sipa_plot <- function(x_lim, beta_lim = c(-1,1) ){
tracks(
    tuQTL = isoform_plot(mylocus = "SIPA1L2", mygene = "SIPA1L2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.9, gwas = "Nalls23andMe_2019", mylim = beta_lim),
  #eQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  #tuQTL = isoform_plot(mylocus = "SIPA1L2", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  #tuQTL = isoform_plot(mylocus = "SIPA1L2", transcript_type = "union", level = "transcript", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  leafcutter = lc_plot("SIPA1L2", type = "leafcutter", PP4 = 0.9, gwas = "Nalls23andMe_2019", mylim = beta_lim),
  SE = junction_plot("SIPA1L2", type = "SE", PP4 = 0.8, gwas = "Nalls23andMe_2019",mylim = beta_lim ),
  A5 = junction_plot("SIPA1L2", type = "A5", PP4 = 0.8, gwas = "Nalls23andMe_2019", flank = 1, mylim =beta_lim ),
  #"QTL SNP" = snp_plot( mylocus = "SIPA1L2",type = "QTL", snp_width = 1, PP4 = 0.8),
   # "GWAS SNP" = snp_plot( mylocus = "SIPA1L2",type = "GWAS", snp_width = 1, PP4 = 0.8),
  "GWAS" = sipa_gwas_plot,
  "SUSIE" = sipa_finemap_plot,
xlim = x_lim, heights = c(1,1,1,1,1,1)
  )
}

sipa_full <- sipa_plot(c(232390122,232642315) )

sipa_mid <- sipa_plot(c(232507037,232642315))

sipa_superzoom <- sipa_plot(c(232533450,232533623))

sipa_closeup <- sipa_plot(c(232500000, 232650000))


ggsave(plot = sipa_full, filename = "plots/SIPA1L2_coloc_plot_full.pdf", width = 7, height = 5 )

ggsave(plot = sipa_mid, filename = "plots/SIPA1L2_coloc_plot_mid.pdf", width = 7, height = 5 )

ggsave(plot = sipa_closeup, filename = "plots/SIPA1L2_coloc_plot_closeup.pdf", width = 7, height = 5 )

sipa_coloc <- filter(all_coloc, locus == "SIPA1L2", PP.H4.abf > 0.5)

sipa_beta_plot <- beta_plot("SIPA1L2",PP4 = 0.8)

ggsave(plot = sipa_beta, filename = "plots/SIPA1L2_coloc_beta_plot.pdf", width = 7, height = 5 )


```

finemapped SNP - "rs3902818"

SIPA1L2 - Leafcutter suggests a cassette exon just upstream of the start codon - isn't predicted to alter reading frame.

Lead QTL SNP for leafcutter cassette exon is chr1:232533485 - 5bp downstream of the 5' splice site!

Plot transcript TPMs against lead SNP

```{r}
sipa_vcf <- read.table("data/SIPA1L2/roussos_v2_unfiltered_SIPA1L2_genotypes.vcf",header=TRUE) %>%
    select(-INFO, -FORMAT, -FILTER, -QUAL, -CHROM, -POS) %>%
  pivot_longer(names_to = "donor", values_to = "value", cols = !c(ID,REF,ALT) ) %>%
  mutate(genotype = case_when(
    grepl("0\\|0", value) ~ paste0(REF, "/",REF),
    grepl("0\\|1|1\\|0", value) ~ paste0(REF, "/",ALT),
    grepl("1|1", value) ~ paste0(ALT, "/",ALT)
   ) )   %>%
  filter(ID == c("rs10797576"), donor %in% roussos_sk$participant_id) %>%
  mutate(genotype = factor(genotype,
                            levels = c(
                              paste0(unique(REF),"/", unique(REF) ),
                              paste0(unique(REF), "/", unique(ALT) ),
                              paste0(unique(ALT),"/",unique(ALT) )
                              )
                            )
  )

sipa_tx <- read_tsv("data/SIPA1L2/roussos_sipa1l2_transcript_tpm.tsv")

# individual TPMs
sipa_tx[c("transcript_id", roussos_sk$sample_id)] %>%
  pivot_longer(names_to = "sample_id", values_to = "TPM", cols = !c(transcript_id)) %>%
  left_join(roussos_sk, by = "sample_id") %>%
  left_join(sipa_vcf, by = c("participant_id" = "donor")) %>%
  filter(!is.na(genotype)) %>%
  ggplot(aes(x = genotype, y = TPM)) + geom_jitter(width = 0.2, height = 0) + geom_boxplot(fill = NA, outlier.colour = NA) +
facet_wrap(~transcript_id, scales = "free") +
  ggpubr::stat_compare_means()
  
# relative proportions
sipa_tu <- sweep(sipa_tx[,-1], MARGIN = 2, STATS = colSums(sipa_tx[,-1]), FUN = "/")
sipa_tu$transcript_id <- sipa_tx$transcript_id

sipa_tu[c("transcript_id", roussos_sk$sample_id)] %>%
  pivot_longer(names_to = "sample_id", values_to = "TPM", cols = !c(transcript_id)) %>%
  left_join(roussos_sk, by = "sample_id") %>%
  left_join(sipa_vcf, by = c("participant_id" = "donor")) %>%
  ggplot(aes(x = genotype, y = TPM)) + geom_jitter(width = 0.2, height = 0) + geom_boxplot(fill = NA, outlier.colour = NA) +
facet_wrap(~transcript_id, scales = "free") +
  ggpubr::stat_compare_means()
  
# manually make a SUPPA-like PSI 
#filter(sipa_tx, transcript_id %in% c("ENST00000674749.1", "ENST00000675407.1") ) %>%
sipa_tx %>% 
  pivot_longer(names_to = "sample_id", values_to = "TPM", cols = !c(transcript_id)) %>%
  left_join(roussos_sk, by = "sample_id") %>%
  left_join(sipa_vcf, by = c("participant_id" = "donor")) %>%
  pivot_wider(names_from = "transcript_id", values_from = "TPM") %>%
  filter(!is.na(genotype)) %>%
  mutate(PSI = (ENST00000674749.1 + MSTRG.6387.2 + MSTRG.6387.16) /  # tx including UTR exon 4
           ( ENST00000674749.1 + MSTRG.6387.2 + MSTRG.6387.16 + 
               MSTRG.6387.27 + MSTRG.6387.6 + ENST00000675407.1 + ENST00000674801.1  )    
         )  %>% # tx including exon 3 but not exon 4
  ggplot(aes(x = genotype, y = PSI, group = genotype)) + 
  geom_violin(scale = "width", fill = "skyblue") +
  geom_jitter(width = 0.1, height = 0) +
  geom_boxplot(width = 0.25, fill =NA, colour = "white",outlier.colour = NA) +
  #sggpubr::stat_compare_means() +
  labs(x = "rs10797576", y = "% Exon inclusion", title = "SIPA1L2 exon 4") +
  theme_classic()



```



## MED12L

```{r}
beta_plot("MED12L", PP4 = 0.8)
tracks(
    tuQTL = isoform_plot(mylocus = "MED12L", transcript_type = "union", level = "transcript", PP4 = 0.5, gwas = "Nalls23andMe_2019"),
    "QTL SNP" = snp_plot( mylocus = "MED12L",type = "QTL", snp_width = 1, PP4 = 0.8),
    "GWAS SNP" = snp_plot( mylocus = "MED12L",type = "GWAS", snp_width = 1, PP4 = 0.8)

)

```


## BST1

```{r}
beta_plot("BST1", PP4 = 0.5, mylim = c(-2,2) )
tracks(
  #eQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  tuQTL = isoform_plot(mylocus = "BST1", transcript_type = "GENCODE", level = "transcript", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
    tuQTL = isoform_plot(mylocus = "BST1", transcript_type = "union", level = "transcript", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  leafcutter = junction_plot("BST1", type = "leafcutter", PP4 = 0.8, gwas = "Nalls23andMe_2019"),
  "QTL SNP" = snp_plot( mylocus = "BST1",type = "QTL", snp_width = 1, PP4 = 0.8),
    "GWAS SNP" = snp_plot( mylocus = "BST1",type = "GWAS", snp_width = 1, PP4 = 0.8)
  )#,
```

No Betas!

## RAB29

```{r}

beta_plot("NUCKS1", PP4 = 0.5)
tracks(
  #eQTL = isoform_plot(mylocus = "SIGLEC11", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim),
  tuQTL = isoform_plot(mylocus = "NUCKS1", transcript_type = "GENCODE", level = "transcript", PP4 = 0.5, gwas = "Nalls23andMe_2019"),
  #tuQTL = isoform_plot(mylocus = "NUCKS1", transcript_type = "union", level = "transcript", PP4 = 0.5, gwas = "Nalls23andMe_2019"),
  leafcutter = junction_plot("NUCKS1", type = "leafcutter", PP4 = 0.5, gwas = "Nalls23andMe_2019"),
    leafcutter = junction_plot("NUCKS1", type = "A5", PP4 = 0.5, gwas = "Nalls23andMe_2019"),
  "QTL SNP" = snp_plot( mylocus = "NUCKS1",type = "QTL", snp_width = 1, PP4 = 0.8),
    "GWAS SNP" = snp_plot( mylocus = "NUCKS1",type = "GWAS", snp_width = 1, PP4 = 0.8)
  )#,
```

FCGR2A

```{r}
sipa_finemap <- read_csv(here::here("external/Schilder_PD_finemapping/FCGR2A.Nalls23andMe_2019.UKB.multi_finemap.csv") )
sipa_coords <- read_tsv(here::here("external/Schilder_PD_finemapping/SIPA1L2.Nalls23andMe_2019.UKB.multi_finemap.hg38.bed"), col_names = c("chr", "start", "end", "SNP", "X") )

sipa_finemap <- left_join(sipa_finemap, sipa_coords, by = "SNP")

sipa_gwas_plot <- sipa_finemap %>%
  ggplot(aes(x = end , y = -log10(P) )) + 
  geom_point(aes(colour = r2), size = 0.5 ) +
  geom_point(data = filter(sipa_finemap, leadSNP == TRUE),  colour = "black", fill = "purple", shape = 23, size = 2) +
  theme_classic() +
  scale_colour_distiller(palette = "RdBu") + 
  labs(colour = "LD R2", x = "position", y = expression(log[10]~P) )

sipa_finemap_plot <-
  sipa_finemap %>%
  ggplot(aes(x = end , y = SUSIE.PP )) + 
  geom_point(aes(colour = 1), size = 1 ) +
  labs(x = "position", y = "PIP") +
  theme_bw() 

sipa_gwas_plot + sipa_finemap_plot +
  plot_layout(ncol = 1)


beta_plot("FCGR2A", PP4 = 0.5)

tracks(
eQTL = isoform_plot(mylocus = "FCGR2A",  level = "expression", PP4 = 0.9, gwas = "Nalls23andMe_2019"),
leafcutter = lc_plot("FCGR2A", type = "leafcutter", PP4 = 0.9, gwas = "Nalls23andMe_2019"),
"GWAS SNP" = snp_plot( mylocus = "FCGR2A",type = "GWAS", snp_width = 1, PP4 = 0.9),
"QTL SNPs" = snp_plot( mylocus = "FCGR2A",type = "QTL", snp_width = 1, PP4 = 0.9),
xlim = c(161492242, 161596553)
)
```


# SCZ / ALS

## CAMLG 

```{r}
beta_plot("locus_5_rs113661575", PP4 = 0.5)
beta_plot("C5orf24", PP4 = 0.5)

siglec11_plot <- function(x_lim, beta_lim = c(-0.6,0.6)){
  tracks(
  eQTL = isoform_plot(mylocus = "locus_5_rs113661575", transcript_type = "GENCODE", level = "expression", PP4 = 0.8, mylim = beta_lim, gwas = "VanRheenenEUR_2021"),
  tuQTL = isoform_plot(mylocus = "locus_5_rs113661575", transcript_type = "isoseq", level = "transcript", PP4 = 0.8, mylim = beta_lim, gwas = "VanRheenenEUR_2021"),
  SE = junction_plot("locus_5_rs113661575", type = "SE", PP4 = 0.8, flank = 10, gwas = "VanRheenenEUR_2021" ),
    leafcutter = junction_plot("locus_5_rs113661575", type = "leafcutter", PP4 = 0.8, flank = 10, gwas = "VanRheenenEUR_2021" ),

  "QTL SNPs" = snp_plot( mylocus = "locus_5_rs113661575",type = "QTL", snp_width = 1, PP4 = 0.8 ),
  "GWAS SNPs" = finemap_plot("locus_5_rs113661575", label = TRUE, min_pip = 0.01, xlim = x_lim, col_width = 10),
  xlim = x_lim
)
}


```


